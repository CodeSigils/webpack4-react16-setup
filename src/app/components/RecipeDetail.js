import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
/**
 * We can pass className to props with expressions like this:
 * `h3 p2 bg-white italic center ${props.className}` but there is the risk
 * of a null value in our html. The classNames library will check if a CSS
 * class is null and will not render it in html, so the above expression
 * can become: classNames('h3 p2 bg-white italic center', props.className)
 */

const RecipeDetail = ({ recipe, style, className }) => {
  /**
   * Handle 0 state of component.
   * If there is no default value React will return a null complain:
   * "cannot read property of null".
   */
  if (!recipe) {
    return (
      <span
        style={style}
        className={classNames('h4 p2 bg-white italic center', className)}
      >
        Select a recipe from the list
      </span>
    );
  }

  return (
    <div style={style} className={classNames('p2 bg-white', className)}>
      <h2 className="h2">{recipe.name}</h2>
      <img className="fit" alt={recipe.name} src={recipe.image} />
      <div>
        <span>{recipe.category}</span>
        <span>{recipe.calories} cal</span>
      </div>
      <h3>Ingredients</h3>
      <ul>
        {recipe.ingredients.map(ingredient => (
          <li key={ingredient}>{ingredient}</li>
        ))}
      </ul>
      <h3>Steps</h3>
      <ol>{recipe.steps.map(step => <li key={step}>{step}</li>)}</ol>
    </div>
  );
};

RecipeDetail.propTypes = {
  recipe: PropTypes.object,
  className: PropTypes.string,
  style: PropTypes.object,
};

export default RecipeDetail;
