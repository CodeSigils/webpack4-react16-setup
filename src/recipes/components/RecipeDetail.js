import React from 'react';
import { PropTypes } from 'prop-types';
import classNames from 'classnames';
import { Link } from 'react-router-dom';
/**
 * We can pass className to props with expressions like this:
 * `h3 p2 bg-white italic center ${props.className}` but there is the risk
 * of a null value in our html. The classNames library will check if a CSS
 * class is null and will not render it in html, so the above expression
 * can become: classNames('h3 p2 bg-white italic center', props.className)
 */

const Timer = () => {
  const tic = new Date().toLocaleTimeString();
  return <p>{tic}</p>;
};

const RecipeDetail = props => {
  /**
   * Handle 0 state of component.
   * If there is no default value React will return a null complain:
   * "cannot read property of null".
   */
  if (!props.recipe) {
    return (
      <span
        style={props.style}
        className={classNames('h4 p2 bg-white italic center', props.className)}
      >
        Select a recipe from the list and if you like it, <br /> add it to
        favorites by clicking the â˜‘ symbol !
        <hr className="border-none mx4 pt2" />
        <Timer tic={props.tic} />
      </span>
    );
  }

  return (
    <div
      style={props.style}
      className={classNames('p2 bg-white', props.className)}
    >
      <h2 className="h2">{props.recipe.name}</h2>
      <img className="fit" alt={props.recipe.name} src={props.recipe.image} />
      <div>
        <span>{props.recipe.category}</span>
        <span>{props.recipe.calories} cal</span>
      </div>
      <h3>Ingredients</h3>
      <ul>
        {props.recipe.ingredients.map(ingredient => (
          <li key={ingredient}>{ingredient}</li>
        ))}
      </ul>
      <h3>Steps</h3>
      <ol>{props.recipe.steps.map(step => <li key={step}>{step}</li>)}</ol>
      <Link
        className="text-decoration-none"
        href={`/recipe/${props.recipe.id}`}
        to={`/recipe/${props.recipe.id}`}
      >
        <span className="p1 black btn rounded">Read more...</span>
      </Link>
    </div>
  );
};

RecipeDetail.propTypes = {
  tic: PropTypes.string,
  recipe: PropTypes.object,
  className: PropTypes.string,
  style: PropTypes.object,
};

export default RecipeDetail;
