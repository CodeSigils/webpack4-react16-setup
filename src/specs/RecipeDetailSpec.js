import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import RecipeDetail from '../recipes/components/RecipeDetail';
// import { classNames } from 'classnames';

describe('<RecipeDetail/>', () => {
  let testRecipe;

  beforeEach(() => {
    testRecipe = {
      id: 1,
      name: 'Test recipe',
      category: 'Test recipe category',
      ingredients: ['Ing 1', 'Ing 2'],
      steps: ['Step 1', 'Step 2'],
    };
  });

  // 1.
  test('Should render zero state', () => {
    const component = renderer.create(
      <BrowserRouter>
        <RecipeDetail />
      </BrowserRouter>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  // 2.
  test('Should correctly render a recipe', () => {
    const component = renderer.create(
      <BrowserRouter>
        <RecipeDetail recipe={testRecipe} />
      </BrowserRouter>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  // 3.
  test('Should correctly assign a className', () => {
    const component = renderer.create(
      <BrowserRouter>
        <RecipeDetail recipe={testRecipe} className="classname-test" />
      </BrowserRouter>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  // 4.
  test('Should render a recipe without ingredients', () => {
    delete testRecipe.ingredients;
    const component = renderer.create(
      <BrowserRouter>
        <RecipeDetail recipe={testRecipe} />
      </BrowserRouter>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  // 5.
  test('Should render a recipe without steps', () => {
    delete testRecipe.steps;
    const component = renderer.create(
      <BrowserRouter>
        <RecipeDetail recipe={testRecipe} />
      </BrowserRouter>,
    );
    const tree = component.toJSON();

    // Tell Jest to check if tree muches the snapshot
    expect(tree).toMatchSnapshot();
  });
});
