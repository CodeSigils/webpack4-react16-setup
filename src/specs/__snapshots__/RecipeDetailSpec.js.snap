// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeDetail/> Should correctly assign a className 1`] = `
<div
  className="p2 bg-white classname-test"
  style={undefined}
>
  <h2
    className="h2"
  >
    Test recipe
  </h2>
  <img
    alt="Test recipe"
    className="fit"
    src={undefined}
  />
  <div>
    <span>
      Test recipe category
    </span>
    <span>
       cal
    </span>
  </div>
  <h3>
    Ingredients
  </h3>
  <ul>
    <li>
      Ing 1
    </li>
    <li>
      Ing 2
    </li>
  </ul>
  <h3>
    Steps
  </h3>
  <ol>
    <li>
      Step 1
    </li>
    <li>
      Step 2
    </li>
  </ol>
  <a
    className="text-decoration-none"
    href="/recipe/1"
    onClick={[Function]}
  >
    <span
      className="p1 black btn rounded"
    >
      Read more...
    </span>
  </a>
</div>
`;

exports[`<RecipeDetail/> Should correctly render a recipe 1`] = `
<div
  className="p2 bg-white"
  style={undefined}
>
  <h2
    className="h2"
  >
    Test recipe
  </h2>
  <img
    alt="Test recipe"
    className="fit"
    src={undefined}
  />
  <div>
    <span>
      Test recipe category
    </span>
    <span>
       cal
    </span>
  </div>
  <h3>
    Ingredients
  </h3>
  <ul>
    <li>
      Ing 1
    </li>
    <li>
      Ing 2
    </li>
  </ul>
  <h3>
    Steps
  </h3>
  <ol>
    <li>
      Step 1
    </li>
    <li>
      Step 2
    </li>
  </ol>
  <a
    className="text-decoration-none"
    href="/recipe/1"
    onClick={[Function]}
  >
    <span
      className="p1 black btn rounded"
    >
      Read more...
    </span>
  </a>
</div>
`;

exports[`<RecipeDetail/> Should render a recipe without ingredients 1`] = `
<div
  className="p2 bg-white"
  style={undefined}
>
  <h2
    className="h2"
  >
    Test recipe
  </h2>
  <img
    alt="Test recipe"
    className="fit"
    src={undefined}
  />
  <div>
    <span>
      Test recipe category
    </span>
    <span>
       cal
    </span>
  </div>
  <h3>
    Ingredients
  </h3>
  <h3>
    Steps
  </h3>
  <ol>
    <li>
      Step 1
    </li>
    <li>
      Step 2
    </li>
  </ol>
  <a
    className="text-decoration-none"
    href="/recipe/1"
    onClick={[Function]}
  >
    <span
      className="p1 black btn rounded"
    >
      Read more...
    </span>
  </a>
</div>
`;

exports[`<RecipeDetail/> Should render a recipe without steps 1`] = `
<div
  className="p2 bg-white"
  style={undefined}
>
  <h2
    className="h2"
  >
    Test recipe
  </h2>
  <img
    alt="Test recipe"
    className="fit"
    src={undefined}
  />
  <div>
    <span>
      Test recipe category
    </span>
    <span>
       cal
    </span>
  </div>
  <h3>
    Ingredients
  </h3>
  <ul>
    <li>
      Ing 1
    </li>
    <li>
      Ing 2
    </li>
  </ul>
  <h3>
    Steps
  </h3>
  <a
    className="text-decoration-none"
    href="/recipe/1"
    onClick={[Function]}
  >
    <span
      className="p1 black btn rounded"
    >
      Read more...
    </span>
  </a>
</div>
`;

exports[`<RecipeDetail/> Should render zero state 1`] = `
<span
  className="h4 p2 bg-white italic center"
  style={undefined}
>
  Select a recipe from the list and if you like it, 
  <br />
   add it to favorites by clicking the â˜‘ symbol !
  <hr
    className="border-none mx4 pt2"
  />
</span>
`;
